datafusion/common/src/error.rs

```diff
@@ -131,6 +132,7 @@ pub enum DataFusionError {
     /// Errors from either mapping LogicalPlans to/from Substrait plans
     /// or serializing/deserializing protobytes to Substrait plans
     Substrait(String),
+    Diagnostic(Diagnostic, Box<DataFusionError>),
```

Comment:
This is an interesting idea -- one way to think about this is that it adds additional structured information to  `DataFusionError::Context`

If we went with the `DataFusion::Diagnostic` approach, do you think we would be able to deprecate / remove `DataFusionError::Context` in a future release?

Reply:
Thank you! Yes, I think `DataFusion::Diagnostic` can convey a superset of the information that `DataFusion::Context` can. Any wrapping such as:

```rust
let error = DataFusionError::Context(message, Box::new(error));
```

can be converted to:

```rust
let error = DataFusionError::Diagnostic(
	Diagnostic {
		entries: vec![
			DiagnosticEntry {
				span: Span::empty(),
				kind: DiagnosticEntryKind::Error,
				message,
			}
		]
	},
	Box::new(error)
);
```

And of course, we can provide a `DataFusionError::with_simple_diagnostic` function to avoid the boilerplate. At that point, `DataFusion::Context` could be removed.

This also enables progressively adding `Span` information to what was previously simply a string message.